### PR TITLE
Update typings for memoize function

### PR DESCRIPTION
**Pull request by gcarling**

### CODE REVIEW

File: src/memoize.ts [hunk 1]
```diff
@@ -18,19 +18,20 @@
...
  * @returns a memoized version of fn
  */
-export function memoize<FnType extends Function>(
+// tslint:disable:no-any defining it this way is more precise than Function so is still preferable
+export function memoize<FnType extends (...args: any[]) => Promise<any>>(
   fn: FnType,
-  // tslint:disable-next-line:no-any (w/o type for Function args, can't assert a type here)
-  hasher: Function = (arg: any) => arg,
+  // tslint:disable:no-any hasher can return any value that can be used as a map key
+  hasher: (...args: Parameters<FnType>) => any = (...args) => args[0],
```
Comment (gcarling) [inline src/memoize.ts:25]: i have nothing to say just wanted to test this out

File: src/memoize.ts [hunk 2]
```diff
@@ -18,19 +18,20 @@
...
   timeoutMs?: number,
 ): FnType & { reset: FnType; clear: () => void } {
   // tslint:disable:no-any (unfortunately we can't give the FnType any more clarity or it limits
   // what you can do with it)
   const memos: Map<any, { value: any; expiration: number }> = new Map();
   const queues: Map<any, Promise<any>> = new Map();
 
-  const returnFn = ((async (...args: any[]): Promise<any> => {
-    const key: any = hasher(...args);
+  const returnFn = ((async (...args: Parameters<FnType>): Promise<any> => {
```
Comment (ftrimble) [inline src/memoize.ts:33]: you can `Promise<ThenReturn<FnType>>` which could potentially avoid a cast somewhere

File: src/memoize.ts [hunk 3]
```diff
@@ -18,19 +18,20 @@
...
+export function memoize<FnType extends (...args: any[]) => Promise<any>>(
   fn: FnType,
-  // tslint:disable-next-line:no-any (w/o type for Function args, can't assert a type here)
-  hasher: Function = (arg: any) => arg,
+  // tslint:disable:no-any hasher can return any value that can be used as a map key
+  hasher: (...args: Parameters<FnType>) => any = (...args) => args[0],
   timeoutMs?: number,
 ): FnType & { reset: FnType; clear: () => void } {
   // tslint:disable:no-any (unfortunately we can't give the FnType any more clarity or it limits
   // what you can do with it)
```
Comment (ftrimble) [inline src/memoize.ts:29]: think we should be able to remove this generaly now

File: src/memoize.ts [hunk 4]
```diff
@@ -18,19 +18,20 @@
...
   const memos: Map<any, { value: any; expiration: number }> = new Map();
```
Comment (ftrimble) [inline src/memoize.ts:30]: think this is a `Map<ReturnType<typeof hasher>, { value: ThenReturn<FnType>; expiration: number }`

File: src/memoize.ts [hunk 5]
```diff
@@ -18,19 +18,20 @@
...
   const queues: Map<any, Promise<any>> = new Map();
```
Comment (ftrimble) [inline src/memoize.ts:31]: `Map<ReturnType<typeof hasher>, Promise<ThenReturn<FnType>>>`

File: src/memoize.ts [hunk 6]
```diff
@@ -18,19 +18,20 @@
...
-export function memoize<FnType extends Function>(
+// tslint:disable:no-any defining it this way is more precise than Function so is still preferable
+export function memoize<FnType extends (...args: any[]) => Promise<any>>(
   fn: FnType,
-  // tslint:disable-next-line:no-any (w/o type for Function args, can't assert a type here)
-  hasher: Function = (arg: any) => arg,
+  // tslint:disable:no-any hasher can return any value that can be used as a map key
+  hasher: (...args: Parameters<FnType>) => any = (...args) => args[0],
   timeoutMs?: number,
 ): FnType & { reset: FnType; clear: () => void } {
```
Comment (ftrimble) [inline src/memoize.ts:27]: think this is actually a `(...args: Parameters<FnType>) => void` (not a promise so the signature is actually wrong for consumers rn)

File: src/memoize.ts [hunk 7]
```diff
@@ -41,19 +42,19 @@ export function memoize<FnType extends Function>(
...
 
     try {
-      const ret: any = await queues.get(key)!;
+      const ret = await queues.get(key)!;
       memos.set(key, { value: ret, expiration: Date.now() + (timeoutMs || 0) });
       return ret;
     } finally {
       queues.delete(key);
     }
   }) as any) as FnType;
```
Comment (ftrimble) [inline src/memoize.ts:55]: shouldn't need this cast anymore

File: test/memoize.test.ts [hunk 8]
```diff
@@ -40,7 +40,7 @@ test('typescript def works for multiple args', async t => {
     return false;
   }
 
-  const memoized = promiseUtils.memoize(stuffToMemoize);
+  const memoized = promiseUtils.memoize(stuffToMemoize, (a: string) => a);
```
Comment (ftrimble) [inline test/memoize.test.ts:43]: was this necessary or just a test?

**Comment by ftrimble** [comment]
just need to `npm version minor`

**Comment by ftrimble** [comment]
also your tests aren't passing

**Pull request merged**